gpytorch/utils/contour_integral_quad.py

```diff
@@ -0,0 +1,137 @@
+import math
+import warnings
+
+import torch
+
+from .broadcasting import _mul_broadcast_shape
+from .linear_cg import linear_cg
+from .minres import minres
+from .warnings import NumericalWarning
+
+
+def contour_integral_quad(
+    lazy_tensor,
+    rhs,
+    inverse=False,
+    weights=None,
+    shifts=None,
+    max_lanczos_iter=20,
+    num_contour_quadrature=7,
+    shift_offset=0,
+):
+    r"""
+    Performs :math:`\mathbf K^{1/2} \mathbf b` or `\mathbf K^{-1/2} \mathbf b`
+    using contour integral quadrature.
+
+    .. note::
+        Right now this only works for non-batch matrices
+
+    :param gpytorch.lazy.LazyTensor lazy_tensor: LazyTensor representing :math:`\mathbf K`
+    :param torch.Tensor rhs: Right hand side tensor :math:`\mathbf b`
+    :param bool inverse: (default False) whether to compute :math:`\mathbf K^{1/2} \mathbf b` (if False)
+        or `\mathbf K^{-1/2} \mathbf b` (if True)
+    :param int max_lanczos_iter: (default 10) Number of Lanczos iterations to run (to estimate eigenvalues)
+    :param int num_contour_quadrature: (default 15) How many quadrature samples to use for approximation
+    :rtype: torch.Tensor
+    :return: Approximation to :math:`\mathbf K^{1/2} \mathbf b` or :math:`\mathbf K^{-1/2} \mathbf b`.
+    """
+    import numpy as np
+    from scipy.special import ellipj, ellipk
```

Comment:
i take it this is already a gpytorch requirement?

Reply:
Actually I looked and we explicitly don't require you to have numpy / scipy installed anywhere and have by convention imported them directly in functions/features where they are used (the same way we handle pyro). I ended up moving the imports down here to follow this convention